File: lib/contrek/finder/concurrent/finder.rb
module Contrek
  module Concurrent
    class Finder
      prepend Poolable

      attr_reader :maximum_width

      # Supported options
      # - number_of_threads: number of threads that can be used by the process. If set to 0, it works in
      #   single-thread mode.
      #   The process can take advantage of multiprocessing when computing the initial contours of the
      #   tiles into which the image is divided, and during the subsequent merging of tiles into pairs.
      #   Each merge is performed by any available thread as soon as two adjacent tiles have been computed.
      #   If set to nil, the process will use the maximum number of cpu available cores.
      # - bitmap: PngBitmap containing the image.
      # - matcher: specific Matcher used for pixel identification.
      # - options: options include:
      #            - number_of_tiles: number of tiles into which the image is initially divided.
      #              It cannot exceed the image width, so the system works with tiles at least one pixel wide.
      #              Given the technique on which the process is based, using excessively narrow tiles is
      #              discouraged.

      def initialize(bitmap:, matcher:, options: {}, &block)
        @initialize_time = Benchmark.measure do
          @block = block
          @tiles = Queue.new
          @whole_tile = nil
          @options = options
          @clusters = []
          @maximum_width = bitmap.w
          @number_of_tiles = options[:number_of_tiles] || (raise "number_of_tiles params is needed!")

          cw = @maximum_width.to_f / @number_of_tiles
          raise "One pixel tile width minimum!" if cw < 1.0
          x = 0
          current_versus = options[:versus]
          raise "Define versus!" if current_versus.nil?

          @number_of_tiles.times do |tile_index|
            tile_end_x = (cw * (tile_index + 1)).to_i

            enqueue!(tile_index: tile_index,
              tile_start_x: x,
              tile_end_x: tile_end_x) do |payload|
              finder = ClippedPolygonFinder.new(
                bitmap: bitmap,
                matcher: matcher,
                options: {versus: current_versus, bounds: true},
                start_x: payload[:tile_start_x],
                end_x: payload[:tile_end_x]
              )
              tile = Tile.new(
                finder: self,
                start_x: payload[:tile_start_x],
                end_x: payload[:tile_end_x],
                name: payload[:tile_index].to_s
              )
              tile.initial_process!(finder)
              @tiles << tile
            end

            x = tile_end_x - 1
          end
          process_tiles!(bitmap)
        end.real
      end

      def process_info(bitmap = nil)
        raw_polygons = @whole_tile.to_raw_polygons

        compress_time = Benchmark.measure do
          if @options.has_key?(:compress)
            FakeCluster.new(raw_polygons, @options).compress_coords
          end
        end.real

        metadata = {
          groups: raw_polygons.size,
          benchmarks: {
            total: ((@initialize_time + compress_time) * 1000).round(3),
            init: (@initialize_time * 1000).round(3),
            outer: (@whole_tile.benchmarks[:outer] * 1000).round(3),
            inner: (@whole_tile.benchmarks[:inner] * 1000).round(3),
            compress: ((compress_time * 1000).round(3) if @options.has_key?(:compress))
          }.compact
        }
        Contrek::Finder::Result.new(raw_polygons, metadata)
      end

      private

      def process_tiles!(bitmap)
        arriving_tiles = []
        loop do
          tile = @tiles.pop
          if tile.whole?
            @whole_tile = tile
            return
          end
          if (twin_tile = arriving_tiles.find { |b| (b.start_x == (tile.end_x - 1)) || ((b.end_x - 1) == tile.start_x) })
            cluster = Cluster.new(finder: self, height: bitmap.h, width: bitmap.w)
            if twin_tile.start_x == (tile.end_x - 1)
              cluster.add(tile)
              cluster.add(twin_tile)
            else
              cluster.add(twin_tile)
              cluster.add(tile)
            end
            enqueue!(cluster: cluster) do |payload|
              merged_tile = payload[:cluster].merge_tiles!
              @tiles << merged_tile
              # usefull external access to each merged_tile
              @block&.call(merged_tile, bitmap)
            end
            arriving_tiles.delete(twin_tile)
            next
          end
          arriving_tiles << tile
        end
      end
    end
  end
end

File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/finder/concurrent/Finder.cpp
/*
 * Finder.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */

#include <utility>
#include <vector>
#include <string>
#include <algorithm>
#include "Finder.h"
#include "../../bitmaps/Bitmap.h"
#include "../../matchers/Matcher.h"
#include "../../matchers/RGBMatcher.h"
#include "../FinderUtils.h"
#include "ClippedPolygonFinder.h"
#include "Tile.h"
#include "Queue.h"
#include "Cluster.h"
#include "FakeCluster.h"

Finder::Finder(int number_of_threads, Bitmap *bitmap, Matcher *matcher, const Options& options)
: Poolable(number_of_threads),
  bitmap(bitmap),
  matcher(matcher),
  input_options(options),
  maximum_width_(bitmap->w()),
  height(bitmap->h())
{ cpu_timer.start();
  FinderUtils::sanitize_options(this->options_, options);

  double cw = static_cast<double>(this->maximum_width_) / this->options_.number_of_tiles;
  if (cw < 1.0) {
    throw std::runtime_error("One pixel tile width minimum!");
  }
  if (this->options_.deterministic && (this->options_.number_of_tiles % 2 != 0)) {
    throw std::invalid_argument("Deterministic mode requires an even number of tiles!");
  }
  int x = 0;
  for (int tile_index = 0; tile_index < this->options_.number_of_tiles; tile_index++)
  { int tile_end_x = static_cast<int>(cw * (tile_index + 1));
    TilePayload p { tile_index, x, tile_end_x };
    enqueue(p, [this](const TilePayload& payload) {
      Options base_arguments = {
        {"bounds", true},
        {"versus", Identifier{this->options_.get_alpha_versus()}},
      };
      if (this->options_.connectivity_offset == 1) {
        base_arguments["connectivity"] = 8;
      }
      if (this->options_.treemap) {
        base_arguments["treemap"] = true;
      }

      CpuTimer t;
      t.start();
      auto* finder = new ClippedPolygonFinder(
        this->bitmap,
        this->matcher,
        payload.tile_start_x,
        payload.tile_end_x,
        base_arguments);
      {
        std::lock_guard<std::mutex> lock(finders_mutex);
        finders.push(finder);
      }

      Tile* tile = new Tile(this, payload.tile_start_x, payload.tile_end_x,
                            std::to_string(payload.tile_index), payload.tile_index, Benchmarks {0, 0});
      tile->initial_process(finder);
      tiles_.queue_push(tile);
    });

    x = tile_end_x - 1;
  }
  this->process_tiles(this->options_.deterministic);
  reports["init"] = cpu_timer.stop();
}

Finder::Finder(int number_of_threads, const Options& options)
: Poolable(number_of_threads), bitmap(nullptr), matcher(nullptr), input_options(options), maximum_width_(0) {
  FinderUtils::sanitize_options(this->options_, options);
  reports["init"] = 0;
}

void Finder::process_tiles(bool deterministic) {
  std::vector<Tile*> arriving_tiles;

  while (true) {
    Tile* tile = tiles_.queue_pop();

    if (tile->whole()) {
      this->whole_tile = tile;
      break;
    }

    auto it = std::find_if(
      arriving_tiles.begin(),
      arriving_tiles.end(),
      [&](Tile* t) {
        bool is_adjacent = (t->start_x() == (tile->end_x() - 1)) ||
                           ((t->end_x() - 1) == tile->start_x());
        if (!is_adjacent) return false;
        if (deterministic && !this->last_couple(tile, t)) {
          if (tile->order() != t->order()) return false;
          if (std::min(t->index(), tile->index()) % 2 != 0) return false;
        }
        return true;
      });

    if (it != arriving_tiles.end()) {
      Tile* twin_tile = *it;

      Cluster *cluster = new Cluster(this, this->height);

      if (twin_tile->start_x() == (tile->end_x() - 1)) {
        cluster->add(tile);
        cluster->add(twin_tile);
      } else {
        cluster->add(twin_tile);
        cluster->add(tile);
      }
      arriving_tiles.erase(it);
      enqueue(cluster, [this](Cluster* c) {
        Tile* merged_tile = c->merge_tiles();
        tiles_.queue_push(merged_tile);
        delete c;
      });
    } else {
      arriving_tiles.push_back(tile);
    }
  }
}

bool Finder::last_couple(const Tile* tile_a, const Tile* tile_b) const {
  bool a_is_first = tile_a->index() < tile_b->index();
  const Tile* first = a_is_first ? tile_a : tile_b;
  const Tile* last  = a_is_first ? tile_b : tile_a;
  return (first->start_x() == 0) && (last->end_x() == this->maximum_width());
}

Finder::~Finder() {
  if (this->whole_tile) {
    delete this->whole_tile;  // last tile to be deleted (not owned by a cluster)
  }

  while (!finders.empty()) {
    ClippedPolygonFinder* finder = nullptr;
    { std::lock_guard<std::mutex> lock(finders_mutex);
      finder = finders.front();
      finders.pop();
    }
    delete finder;
  }
}

ProcessResult* Finder::process_info() {
  ProcessResult *pr = new ProcessResult();
  pr->polygons = std::move(this->whole_tile->to_raw_polygons());
  pr->groups = pr->polygons.size();
  pr->width = this->maximum_width_;
  pr->height = this->height;
  pr->has_bounds = this->options_.bounds;
  pr->versus = this->options_.versus;
  pr->options = this->input_options;
  FakeCluster fake_cluster(pr->polygons, this->options_);
  cpu_timer.start();
  fake_cluster.compress_coords(pr->polygons, this->options_);
  reports["compress"] = cpu_timer.stop();
  reports["total"] = reports["compress"] + reports["init"];
  reports["outer"] = this->whole_tile->benchmarks.outer;
  reports["inner"] = this->whole_tile->benchmarks.inner;
  if (this->options_.treemap) {
    pr->treemap = this->whole_tile->compute_treemap();
  }
  pr->benchmarks = this->reports;
  // TODO(ema): pr->named_sequence
  return(pr);
}
